File: typescript/problem_10.ts
// schedule calls f after n milliseconds using setTimeout.
export function schedule(f: () => void, n: number): void {
    setTimeout(f, n);
}

export async function problem10(): Promise<void> {
    console.log("Problem 10:");
    await new Promise<void>((resolve) => schedule(resolve, 50));
    console.log("All Tests Passed");
}

File: go/problem_10.go
package main

import (
	"fmt"
	"time"
)

// schedule calls f after a delay of n milliseconds in a background goroutine.
func schedule(f func(), n int) {
	go func() {
		time.Sleep(time.Duration(n) * time.Millisecond)
		f()
	}()
}

func problem10() {
	fmt.Println("Problem 10:")
	done := make(chan struct{})
	schedule(func() {
		close(done)
	}, 50)
	<-done // block until the scheduled function fires
	fmt.Println("All Tests Passed")
}
